### PR TITLE
AWS Lambda integration : Add wavefront python wrapper

### PR DESCRIPTION
**Pull request by akodali18**
Add wavefront python wrapper to report standard lambda metrics directly to wavefront.

This wrapper will enable aws lambda functions to report standard lambda function metrics as well as business metrics directly to wavefront bypassing cloudwatch metrics.

Standard lambda function metrics collected are:
    1.Lambda Invocations
    2.Lambda Duration
    3.Lambda Errors
    4.Lambda cold starts

### CODE REVIEW

File: wavefront_lambda/README.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+## Environmental variables
+WAVEFRONT_URL = you corp wavefront url [Ex: https://yourcorp.wavefront.com]  
+WAVEFRONT_API_TOKEN = your wavefront api authentication token  
+IS_REPORT_STANDARD_METRICS = set to false to not report standard lambda metrics directly to wavefront.  
+
+## Usage
+
+Decorate your AWS Lambda handler function with @wavefront_lambda.wrapper.
+
+You can create a `WavefrontProxyReporter` or `WavefrontDirectReporter`:
```
Comment (vikramraman) [inline wavefront_lambda/README.md]: Remove this line?

Comment (akodali18) [reply wavefront_lambda/README.md]: Done

File: wavefront_lambda/README.md [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+## Installation
+To install from PyPi
+```
+pip install wavefront_lambda
+```
+
+## Environmental variables
+WAVEFRONT_URL = you corp wavefront url [Ex: https://yourcorp.wavefront.com]  
+WAVEFRONT_API_TOKEN = your wavefront api authentication token
```
Comment (vikramraman) [inline wavefront_lambda/README.md]: For URL, let's just say: `https://<INSTANCE>.wavefront.com`.
For token: Wavefront API token with Direct Data Ingestion permission.

File: wavefront_lambda/wavefront_lambda/__init__.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+reg = None
+
+
+def wrapper(func):
+    """
+    Returns the Wavefront Lambda wrapper. The wrapper collects aws lambda's
+    standard metrics and reports it directly to the set wavefront url. It
+    requires the following Environmental variables to be set:
+    1.WAVEFRONT_URL : your corp wavefront url
+    2.WAVEFRONT_API_TOKEN : The users API authentication token
```
Comment (vikramraman) [inline wavefront_lambda/wavefront_lambda/__init__.py]: Same as above, mention Direct Data Ingestion permission is required for the associated token.

File: wavefront_lambda/wavefront_lambda/__init__.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+is_cold_start = True
+reg = None
+
+
+def wrapper(func):
+    """
+    Returns the Wavefront Lambda wrapper. The wrapper collects aws lambda's
+    standard metrics and reports it directly to the set wavefront url. It
+    requires the following Environmental variables to be set:
+    1.WAVEFRONT_URL : your corp wavefront url
```
Comment (vikramraman) [inline wavefront_lambda/wavefront_lambda/__init__.py]: Mention Wavefront URL of the form `https://<INSTANCE>.wavefront.com`

File: wavefront_lambda/wavefront_lambda/__init__.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+        try:
+            result = func(*args, **kwargs)
+            return result
+        except:
+            raise
+        finally:
+            wf_reporter.report_now(registry=reg)
+
+    def wavefront_wrapper(*args, **kwargs):
+        auth_token = os.environ.get('WAVEFRONT_API_TOKEN')
```
Comment (vikramraman) [inline wavefront_lambda/wavefront_lambda/__init__.py]: check if token, server is non-empty?

File: wavefront_lambda/wavefront_lambda/__init__.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+        global reg
+        reg = MetricsRegistry()
+
+        # Initialize the wavefront direct reporter
+        wf_direct_reporter = WavefrontDirectReporter(server=server,
+                                                     token=auth_token,
+                                                     registry=reg,
+                                                     source=point_tags['FunctionName'],
+                                                     tags=point_tags,
+                                                     prefix="wf.aws.lambda.")
```
Comment (vikramraman) [inline wavefront_lambda/wavefront_lambda/__init__.py]: let's go with `aws.lambda.wf` as the prefix.

File: wavefront_lambda/wavefront_lambda/__init__.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+    2.WAVEFRONT_API_TOKEN : The users API authentication token
+    3.IS_REPORT_STANDARD_METRICS : set to false to not report standard lambda
+                                   metrics directly to wavefront.
+
+    """
+    def call_lambda_with_standard_metrics(wf_reporter, *args, **kwargs):
+        # Set cold start counter
+        global is_cold_start
+        if is_cold_start:
+            aws_cold_starts_counter = delta.delta_counter(reg, "coldstart")
```
Comment (vikramraman) [inline wavefront_lambda/wavefront_lambda/__init__.py]: Let's rename this to `coldstarts`.

File: wavefront_lambda/wavefront_lambda/__init__.py [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+            aws_cold_starts_counter.inc()
+            aws_cold_starts_normal_counter = reg.counter("coldstart-normalcounter")
```
Comment (vikramraman) [inline wavefront_lambda/wavefront_lambda/__init__.py]: Let's remove the `-normalcounter` suffix. Maybe `coldstarts_raw`?

Also wondering if this should be a gauge, is it possible for this counter to ever be > 1? i.e for two cold starts resulting in the same counter instance being used?

Comment (akodali18) [reply wavefront_lambda/wavefront_lambda/__init__.py]: cold start can only be True/False where True => 1 , If a 2nd container is spun up as a result of concurrent invocations and hence the 2nd cold start , then the 2nd container will use a new registry, so above scenario should not come up.

**Pull request merged**